pkg/controller/clusterset/clusterset.go

```diff
@@ -340,3 +329,100 @@ func (c *ClusterSetReconciler) reconcileClusters(ctx context.Context, clusterSet
 
 	return err
 }
+
+func (c *ClusterSetReconciler) reconcileQuota(ctx context.Context, clusterSet *v1alpha1.ClusterSet) error {
+	if clusterSet.Spec.Quota == nil {
+		// check if resourceQuota object exists and deletes it.
+		var toDeleteResourceQuota v1.ResourceQuota
+
+		key := types.NamespacedName{
+			Name:      k3kcontroller.SafeConcatNameWithPrefix(clusterSet.Name),
+			Namespace: clusterSet.Namespace,
+		}
+
+		if err := c.Client.Get(ctx, key, &toDeleteResourceQuota); err != nil {
+			return client.IgnoreNotFound(err)
+		}
+
+		return c.Client.Delete(ctx, &toDeleteResourceQuota)
+	}
+
+	// create/update resource Quota
+	resourceQuota := resourceQuota(clusterSet)
+
+	if err := ctrl.SetControllerReference(clusterSet, &resourceQuota, c.Scheme); err != nil {
+		return err
+	}
+
+	if err := c.Client.Create(ctx, &resourceQuota); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return c.Client.Update(ctx, &resourceQuota)
+		}
+	}
+
+	return nil
+}
+
+func resourceQuota(clusterSet *v1alpha1.ClusterSet) v1.ResourceQuota {
+	return v1.ResourceQuota{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      k3kcontroller.SafeConcatNameWithPrefix(clusterSet.Name),
+			Namespace: clusterSet.Namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ResourceQuota",
+			APIVersion: "v1",
+		},
+		Spec: v1.ResourceQuotaSpec{
+			Hard:          clusterSet.Spec.Quota.Limits,
+			Scopes:        clusterSet.Spec.Quota.Scopes,
+			ScopeSelector: clusterSet.Spec.Quota.ScopeSelector,
+		},
+	}
+}
+
+func (c *ClusterSetReconciler) reconcileDefaultLimits(ctx context.Context, clusterSet *v1alpha1.ClusterSet) error {
+	limitRange := limitRange(clusterSet)
+	if err := ctrl.SetControllerReference(clusterSet, &limitRange, c.Scheme); err != nil {
+		return err
+	}
+
+	if err := c.Client.Create(ctx, &limitRange); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return c.Client.Update(ctx, &limitRange)
+		}
+	}
+
+	return nil
+}
+
+func limitRange(clusterSet *v1alpha1.ClusterSet) v1.LimitRange {
+	items := clusterSet.Spec.DefaultLimits
+	if items == nil {
+		// if defaultlimits is not specified we attempt to have a default value for
+		// requests for cpu/memory for the pods within the clusterset
+		items = []v1.LimitRangeItem{
+			{
+				Type: v1.LimitTypeContainer,
+				DefaultRequest: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse(defaultCPURequest),
+					v1.ResourceMemory: resource.MustParse(defaultMemoryRequest),
+				},
+			},
+		}
+	}
```

Comment:
Hm, I have the feeling that this check could be overcome. We are setting a default value only if the `clusterSet.Spec.DefaultLimits` is nil, but since this is now a `[]v1.LimitRangeItem` then the user could set a different type if `LimitRangeItem` (i.e. if he set just a LimitTypePersistentVolumeClaim), and he will have an unbound limit for the containers.

I think I'm fine with this, maybe we should not provide a default at all, and let the users handle their limits. Obviously documenting the behavior.

I think setting defaults could sometimes being kind of unexpected. Not sure.

Reply:
The issue is that without a default limit the resourceQuota will not work, because Memory/CPU must be specified for all pods within a resourceQuota's namespace:

From Docs:

> If quotas are enabled in a namespace for compute resources like cpu and memory, users must specify requests or limits for those values; otherwise, the quota system may reject pod creation. Hint: Use the LimitRanger admission controller to force defaults for pods that make no compute resource requirements.